### PR TITLE
Featureimage/ipmprog 1244 discovery migration

### PR DESCRIPTION
**Pull request by perrettecl**

### CODE REVIEW

**Comment by etnKlendathu** [comment]
where is difference with previous PR (https://github.com/42ity/fty-pack/pull/24)?

File: tests/map.cpp [hunk 1]
```diff
@@ -177,7 +179,7 @@ TEST_CASE("Simple map serialization/deserialization")
     origin.ints.append("key1", 12);
     origin.ints.append("key2", 13);
 
-    auto check = [](const TestMap& item) {
+    auto check = [](TestMap& item) {
```
Comment (etnKlendathu) [inline tests/map.cpp:182]: should be const

File: tests/map.cpp [hunk 2]
```diff
@@ -230,25 +232,73 @@ TEST_CASE("Simple map serialization/deserialization")
...
+    CHECK(origin["key1"].value == "Some string modified");
+    origin.set("key1", sav);
+    CHECK(origin["key1"].value == "Some string");
 
     MapObj ins;
     ins.value = "Some other value";
     origin.append("key2", ins);
 
-    auto check = [](const pack::Map<MapObj>& item) {
+    auto check = [](pack::Map<MapObj>& item) {
```
Comment (etnKlendathu) [inline tests/map.cpp:301]: should be const

File: pack/map.h [hunk 3]
```diff
@@ -86,26 +91,41 @@ class Map : public IObjectMap
...
     Iterator      begin();
     Iterator      end();
 
+    /// Returns internal std container from the Map
     const MapType& value() const;
+    /// Set values of the map by internal representation
     void           setValue(const MapType& val);
+    /// Returns true if the key exists in the map
     bool           contains(const std::string& key) const;
-    const T&       operator[](const std::string& key) const;
```
Comment (etnKlendathu) [inline pack/map.h:97]: Why did you delete cons variant?

File: pack/map.h [hunk 4]
```diff
@@ -138,22 +162,35 @@ class ValueMap : public IValueMap
...
     Iterator      begin();
     Iterator      end();
 
+    /// Get map
     const MapType& value() const;
+    /// Set values of the map by internal representation
     void           setValue(const MapType& val);
+    /// Returns true if the key exists in the map
     bool           contains(const std::string& key) const;
-    const CppType& operator[](const std::string& key) const;
```
Comment (etnKlendathu) [inline pack/map.h:149]: Why did you delete const variant?

**Review by etnKlendathu — CHANGES_REQUESTED** [review]

**Comment by etnKlendathu** [comment]
Cool :) Just return const in check of object in UT. :) And everything will be perfect. :)

File: tests/map.cpp [hunk 5]
```diff
@@ -263,16 +321,109 @@ TEST_CASE("Object map serialization/deserialization")
...
+
+    auto sav       = origin["key1"];
+    origin["key1"] = "Some string modified";
+    CHECK(origin["key1"] == "Some string modified");
+    origin.set("key1", sav);
+    CHECK(origin["key1"] == "Some string");
+
+    origin.append("key2", "Some other value");
+
+    auto check = [](pack::StringMap& item) {
```
Comment (etnKlendathu) [inline tests/map.cpp:402]: const :)

Comment (eric3873) [reply tests/map.cpp:402]: Yes could be now

File: tests/map.cpp (hunk 2 above)
Comment (perrettecl) [inline tests/map.cpp:309]:
```suggestion
    auto check = [](const pack::Map<MapObj>& item) {
```

File: tests/map.cpp [hunk 6]
```diff
@@ -263,16 +313,100 @@ TEST_CASE("Object map serialization/deserialization")
...
```
Comment (perrettecl) [inline tests/map.cpp:402]:
```suggestion
    auto check = [](const pack::StringMap& item) {
```

File: tests/map.cpp (hunk 1 above)
Comment (perrettecl) [inline tests/map.cpp:182]:
```suggestion
    auto check = [](const TestMap& item) {
```

Comment (eric3873) [reply tests/map.cpp]: Sorry, it has been yet taken into account, the feature has not been updated

**Review by eric3873 — APPROVED** [review]
